apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/ResponseField.kt

```diff
@@ -45,73 +44,21 @@ open class ResponseField internal constructor(
     }
   }
 
-  override fun equals(other: Any?): Boolean {
-    if (this === other) return true
-    if (other !is ResponseField) return false
+  sealed class Type {
+    class NotNull(val ofType: Type): Type()
+    class List(val ofType: Type): Type()
 
-    if (type != other.type) return false
-    if (responseName != other.responseName) return false
-    if (fieldName != other.fieldName) return false
-    if (arguments != other.arguments) return false
-    if (optional != other.optional) return false
-    if (conditions != other.conditions) return false
-
-    return true
-  }
-
-  override fun hashCode(): Int {
-    var result = type.hashCode()
-    result = 31 * result + responseName.hashCode()
-    result = 31 * result + fieldName.hashCode()
-    result = 31 * result + arguments.hashCode()
-    result = 31 * result + optional.hashCode()
-    result = 31 * result + conditions.hashCode()
-    return result
-  }
-
-  /**
-   * An abstraction for the field types
-   */
-  enum class Type {
-    STRING, INT, DOUBLE, BOOLEAN, ENUM, OBJECT, LIST, CUSTOM
-  }
-
-  /**
-   * Abstraction for a Field representing a custom GraphQL scalar type.
-   */
-  class CustomScalarField internal constructor(
-      responseName: String,
-      fieldName: String,
-      arguments: Map<String, Any?>?,
-      optional: Boolean,
-      conditions: List<Condition>?,
-      val customScalar: CustomScalar
-  ) : ResponseField(
-      type = Type.CUSTOM,
-      responseName = responseName,
-      fieldName = fieldName,
-      arguments = arguments.orEmpty(),
-      optional = optional,
-      conditions = conditions.orEmpty()
-  ) {
-
-    override fun equals(other: Any?): Boolean {
-      if (this === other) return true
-      if (other !is CustomScalarField) return false
-      if (!super.equals(other)) return false
-
-      if (customScalar != other.customScalar) return false
-
-      return true
-    }
-
-    override fun hashCode(): Int {
-      var result = super.hashCode()
-      result = 31 * result + customScalar.hashCode()
-      return result
-    }
+    /**
+     * a Named GraphQL type
+     *
+     * @param kind: whether this is an object or not. This is currently required by the CacheKeyResolver API
+     * In a typical server scenario, the resolvers would have access to the schema and would look up the complete typ
+     * but we want to stay lightweight so for now we add this information
+     */
+    class Named(val name: String, val kind: Kind): Type()
   }
 
+  enum class Kind { OBJECT, OTHER }
```

Comment:
nit: Are we expecting to have more than 2 enums ? Should we then just remove it and break sealed class `Named` for 2 classes?

Reply:
Yep, I was wondering about this as well. I'm not 100% sure where this cache resolvers logic will take us. If we take the backend analogy, a server would have access to the schema and would retrieve the actual GraphQL type information from the name only. 
We still want this to be lightweight so it makes sense to embed some schema information directly inside the response field. It just feels weird to do it only for object types. I'm thinking that in the long term, we might want `fromFieldArguments()` to be applicable to something else than just object types and we might need more information stored there.
Also, `NotNull`/`List`/`Named` are taken directly from the GraphQL spec so it maps easily. 

Maybe we could do something like :

```
  sealed class Type {
    class NotNull(val ofType: Type): Type()
    class List(val ofType: Type): Type()

    sealed class Named(val name: String): Type() {
      class Object(name: String): Named(name)
      class Other(name: String): Named(name)
    }
  }
```

Reply:
Yeah I think proposed ^^^ sealed class works great

Reply:
Ooops sorry, I wasn't sure sealed classes could contain other sealed classes, I didn't realize this is what you proposed in the first place. This is done.